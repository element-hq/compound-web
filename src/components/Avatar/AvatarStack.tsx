/*
Copyright 2023  New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { useEffect } from "react";
import { renderToStaticMarkup } from "react-dom/server";
import styles from "./Avatar.module.css";

import AvatarClipPath from "./avatar-clip.mask.svg";

type AvatarStackProps = {
  className?: string;
};

const AVATAR_MASK_ID = "cpdAvatarClipSvg";
let AvatarStackUsageCount = 0;

/**
 * Renders a stack of avatars and clips the content appropriately.
 *
 * The `type` of avatars should always be set to `round`
 * And all the avatars should have the same size.
 */
export const AvatarStack = ({
  children,
  className,
}: React.PropsWithChildren<AvatarStackProps>): React.JSX.Element => {
  /**
   * The `clip-path` property in CSS supports a `path()` function, however
   * that has to use pixel values.
   * `clipPathUnits="objectBoundingBox"` only exists inside an SVG document.
   * Which is why we have the external `avatar-clip.mask.svg`.
   *
   * It is possible to load this SVG externally using `url(path/to/doc.svg#id)`
   * But this is only supported in Firefox
   * This leaves us with no choice but rendering SVG inline in the HTML document
   * And making sure there is always only a single instance regardless of the
   * amount of `AvatarStack` instance in the document.
   */
  useEffect(() => {
    if (AvatarStackUsageCount === 0) {
      const svgMask = renderToStaticMarkup(
        <AvatarClipPath id={AVATAR_MASK_ID} />
      );
      document.body.insertAdjacentHTML("beforeend", svgMask);
    }
    AvatarStackUsageCount++;
    return () => {
      AvatarStackUsageCount--;
      if (AvatarStackUsageCount <= 0) {
        // Cleanup the mask if there are no instances of `AvatarStack` left
        // on the page
        document.getElementById(AVATAR_MASK_ID)?.remove();
      }
    };
  }, []);

  return (
    <div className={classnames(styles["stacked-avatars"], className)}>
      {children}
    </div>
  );
};
