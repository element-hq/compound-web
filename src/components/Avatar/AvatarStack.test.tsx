/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { render } from "@testing-library/react";
import React from "react";

import { AvatarStack } from "./AvatarStack";

const originalImage = global.Image;

describe("AvatarStack", () => {
  beforeEach(() => {
    global.Image = class extends originalImage {
      constructor() {
        super();
        setTimeout(() => {
          this.onload?.(new Event("load"));
        }, 50);
      }
    };
  });

  afterEach(() => {
    global.Image = originalImage;
  });

  const avatars = [
    { id: "@alice:example.org", name: "Alice" },
    { id: "@bob:example.org", name: "Bob" },
  ];

  it("renders", () => {
    const { asFragment } = render(
      <AvatarStack avatars={avatars} size="32px" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("adds the mask to the body", () => {
    render(<AvatarStack avatars={avatars} size="32px" />);

    // We can't run better assertions as `SVG` loading is mocked out in jest
    expect(document.querySelectorAll("svg")).toHaveLength(1);

    render(<AvatarStack avatars={avatars} size="32px" />);

    // We only one instance of the mask, ever!
    expect(document.querySelectorAll("svg")).toHaveLength(1);
  });
});
