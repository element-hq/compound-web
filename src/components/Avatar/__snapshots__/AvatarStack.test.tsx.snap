// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarStack renders 1`] = `
<DocumentFragment>
  <div
    class="stacked-avatars"
  />
</DocumentFragment>
`;
